eventlet/dagpool.py

```diff
@@ -0,0 +1,484 @@
+        # the values.
+        self.wait(self.coros.keys())
+        # As promised, return all the values.
+        return self.values
+
+    def spawn(self, key, depends, function, *args, **kwds):
+        """
+        Launch the passed function(key, results, ...) as a greenthread,
+        passing it:
+
+        - the specified 'key'
+        - an iterable of (key, value) pairs
+        - whatever other positional args or keywords you specify.
+
+        Iterating over the 'results' iterable behaves like calling
+        wait_each(depends).
+
+        Returning from function() behaves like post(key, return_value).
+
+        If you pass spawn() a key already passed to spawn() or post(), spawn()
+        raises Collision.
+        """
+        if key in self.coros or key in self.values:
+            raise Collision(key)
+
+        # The order is a bit tricky. First construct the set() of keys.
+        pending = set(depends)
+        # It's important that we pass to _wait_each() the same 'pending' set()
+        # that we store in self.coros for this key. The generator-iterator
+        # returned by _wait_each() becomes the function's 'results' iterable.
+        newcoro = greenthread.spawn(self._wrapper, function, key,
+                                    self._wait_each(pending),
+                                    *args, **kwds)
+        # Also capture the same (!) set in the new _Coro object for this key.
+        # We must be able to observe ready keys being removed from the set.
+        self.coros[key] = self._Coro(newcoro, pending)
+
+    def _wrapper(self, function, key, results, *args, **kwds):
+        """
+        This wrapper runs the top-level function in a DAGPool greenthread,
+        posting its return value to the DAGPool.
+        """
+        try:
+            # call our passed function
+            result = function(key, results, *args, **kwds)
+        finally:
+            # function() has returned (or raised an exception). We no longer
+            # need to track this greenthread in self.coros. Remove it first so
+            # post() won't complain about a running greenthread.
+            del self.coros[key]
+
+        try:
+            # as advertised, try to post() our return value
+            self.post(key, result)
+        except Collision:
+            # if we've already post()ed a result, oh well
+            pass
+
+        # also, in case anyone cares...
+        return result
+
+    def spawn_many(self, depends, function, *args, **kwds):
+        """
+        spawn_many() accepts a single function whose parameters are the same
+        as for spawn().
+
+        The difference is that spawn_many() accepts a dependency dict. A new
+        greenthread is spawned for each key in the dict. That dict key's value
+        should be an iterable of other keys on which this greenthread depends.
+        """
+        # Iterate over a snapshot of 'depends' items
+        for key, deps in depends.items():
```

Comment:
In Python3 `dict.items()` is not a snapshot, instead it's a view that sees changes in dict.